code/datums/mapgen/JungleGenerator.dm

```diff
@@ -0,0 +1,63 @@
+//the random offset applied to square coordinates, causes intermingling at biome borders
+#define BIOME_RANDOM_SQUARE_DRIFT 2
+
+/datum/map_generator/jungle_generator
+	///2D list of all biomes based on heat and humidity combos.
+	var/list/possible_biomes = list(BIOME_LOW_HEAT = list(BIOME_LOW_HUMIDITY = /datum/biome/plains, BIOME_LOWMEDIUM_HUMIDITY = /datum/biome/mudlands, BIOME_HIGHMEDIUM_HUMIDITY = /datum/biome/mudlands, BIOME_HIGH_HUMIDITY = /datum/biome/water), BIOME_LOWMEDIUM_HEAT = list(BIOME_LOW_HUMIDITY = /datum/biome/plains, BIOME_LOWMEDIUM_HUMIDITY = /datum/biome/jungle, BIOME_HIGHMEDIUM_HUMIDITY = /datum/biome/jungle, BIOME_HIGH_HUMIDITY = /datum/biome/mudlands), BIOME_HIGHMEDIUM_HEAT = list(BIOME_LOW_HUMIDITY = /datum/biome/plains, BIOME_LOWMEDIUM_HUMIDITY = /datum/biome/plains, BIOME_HIGHMEDIUM_HUMIDITY = /datum/biome/jungle/deep, BIOME_HIGH_HUMIDITY = /datum/biome/jungle), BIOME_HIGH_HEAT = list(BIOME_LOW_HUMIDITY = /datum/biome/wasteland, BIOME_LOWMEDIUM_HUMIDITY = /datum/biome/plains, BIOME_HIGHMEDIUM_HUMIDITY = /datum/biome/jungle, BIOME_HIGH_HUMIDITY = /datum/biome/jungle/deep))
+
+///Seeds the rust-g perlin noise with a random number.
+/datum/map_generator/jungle_generator/generate_terrain(var/list/turfs)
+	. = ..()
+	var/height_seed = rand(0, 50000)
+	var/humidity_seed = rand(0, 50000)
+	var/heat_seed = rand(0, 50000)
+
+	for(var/turf/T in turfs) //Go through all the turfs and generate them
+		var/x = T.x + rand(-BIOME_RANDOM_SQUARE_DRIFT, BIOME_RANDOM_SQUARE_DRIFT)
+		var/y = T.y + rand(-BIOME_RANDOM_SQUARE_DRIFT, BIOME_RANDOM_SQUARE_DRIFT)
+
+		var/height = CLAMP01(text2num(rustg_noise_get_at_coordinates("[height_seed]", "[x / 65]", "[y / 65]")))
```

Comment:
What's with these /65

Reply:
I have to make these into defines but basically perlin noise input isn't pure coords as at any integer value it reaches 0. Instead we want something smaller. I divide it by 65 and thus transition between biomes is at a nice speed. If I lower it there'll be faster transitioning, higher and it becomes slower. So that should just be a var or define